contracts/mainnet/connectors/morpho-aave-v3/main.sol

```diff
@@ -0,0 +1,681 @@
+	 */
+	function withdrawWithMaxIterations(
+		address _tokenAddress,
+		uint256 _amount,
+		address _onBehalf,
+		address _receiver,
+		uint256 _maxIteration,
+		uint256 _getId,
+		uint256 _setId
+	)
+		external
+		payable
+		returns (string memory _eventName, bytes memory _eventParam)
+	{
+		uint256 _amt = getUint(_getId, _amount);
+
+		MORPHO_AAVE_V3.withdraw(_tokenAddress, _amt, _onBehalf, _receiver, _maxIteration);
+
+		convertWethToEth(_tokenAddress == ethAddr, TokenInterface(wethAddr), _amt);
+
+		setUint(_setId, _amt);
+
+		_eventName = "LogWithdrawWithMaxIterations(address,uint256,address,uint256,uint256,uint256)";
+		_eventParam = abi.encode(
+			_tokenAddress,
+			_amt,
+			_receiver,
+			_maxIteration,
+			_getId,
+			_setId
+		);
+	}
+
+	function withdrawCollateral(
+		address _tokenAddress,
+		uint256 _amount,
+		address _receiver,
+		uint256 _getId,
+		uint256 _setId
+	)
+		external
+		payable
+		returns (string memory _eventName, bytes memory _eventParam)
+	{
+		uint256 _amt = getUint(_getId, _amount);
+
+		MORPHO_AAVE_V3.withdrawCollateral(_tokenAddress, _amt, address(this), _receiver);
+
+		convertWethToEth(_tokenAddress == ethAddr, TokenInterface(wethAddr), _amt);
+
+		setUint(_setId, _amt);
+
+		_eventName = "LogWithdrawCollateral(address,uint256,address,uint256,uint256)";
+		_eventParam = abi.encode(
+			_tokenAddress,
+			_amt,
+			_receiver,
+			_getId,
+			_setId
+		);
+	}
+
+	function withdrawCollateralOnBehalf(
+		address _tokenAddress,
+		uint256 _amount,
+		address _onBehalf,
+		address _receiver,
+		uint256 _getId,
+		uint256 _setId
+	)
+		external
+		payable
+		returns (string memory _eventName, bytes memory _eventParam)
+	{
+		uint256 _amt = getUint(_getId, _amount);
+		MORPHO_AAVE_V3.withdrawCollateral(_tokenAddress, _amt, _onBehalf, _receiver);
+		convertWethToEth(_tokenAddress == ethAddr, TokenInterface(wethAddr), _amt);
+
+		setUint(_setId, _amt);
+
+		_eventName = "LogWithdrawCollateralOnBehalf(address,uint256,address,address,uint256,uint256)";
+		_eventParam = abi.encode(
+			_tokenAddress,
+			_amt,
+			_onBehalf,
+			_receiver,
+			_getId,
+			_setId
+		);
+	}
+
+	/**
+	 * @dev Payback ETH/ERC20_Token.
+	 * @notice Payback a token to Morpho Aave.
+	 * @param _tokenAddress The address of underlying token to payback.(For ETH: 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE)
+	 * @param _amount The amount of the token (in underlying) to payback. (For max: `uint256(-1)`)
+	 * @param _getId ID to retrieve amt.
+	 * @param _setId ID stores the amount of tokens paid back.
+	 */
+	function payback(
+		address _tokenAddress,
+		uint256 _amount,
+		uint256 _getId,
+		uint256 _setId
+	)
+		external
+		payable
+		returns (string memory _eventName, bytes memory _eventParam)
+	{
+		bool _isETH = _tokenAddress == ethAddr;
+		uint256 _amt = getUint(_getId, _amount);
+
+		TokenInterface _tokenContract = _isETH
+			? TokenInterface(wethAddr)
+			: TokenInterface(_tokenAddress);
+
+		if (_amt == uint256(-1)) {
```

Comment:
Update the logics -
If amount is max =>
1. calculate dsaBalance for token (eth/token)
2. Calculate debt amount using (MORPHO_AAVE_V3.borrowBalance())
3. Allocate _amt as minimum of the 2 values above.